### PR TITLE
HEVC FEI support

### PR DESCRIPTION
**Pull request by xfengcarl**
HEVC FEI support, new HEVC FEI buffer in va.h and add va_fei_hevc.h

Signed-off-by: Kelvin Hu <kelvin.hu@intel.com>
Signed-off-by: xfengcarl <carl.zhang@intel.com>

### CODE REVIEW

**Comment by xfengcarl** [comment]
because it include va_fei.h, so it cant pass build. but va_fei.h is not checked in now, not sure whether I should set different compare target.

**Comment by uartie** [comment]
please rebase as v2.0-branch has va_fei.h now.

File: va/va_fei_hevc.h [hunk 1]
```diff
@@ -0,0 +1,282 @@
...
+    // DWORD 4
+    uint32_t    reserved3                        : 16;  /* sse_classID_32x32_0~3 */
+    uint32_t    lcu_force_zero_coeff             : 1;
+    uint32_t    disable_sao_on_a_lcu_flag        : 1;
+    uint32_t    reserved4                        : 14;
+
+    // DWORD 5~7
+    uint32_t    reserved5[3];  /* padding zero DWORDs */
+} VAEncFEICTBCmdLongHevc;   // 32 bytes
+
```
Comment (xhaihao) [inline va/va_fei_hevc.h]: Missed '}' for __cplusplus

File: va/va_fei_hevc.h [hunk 2]
```diff
@@ -0,0 +1,282 @@
...
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#include <stdint.h>
+#include "va_fei.h"
+
+/** \brief FEI frame level control buffer for Hevc */
+typedef struct _VAEncMiscParameterFEIFrameControlHevc
```
Comment (xhaihao) [inline va/va_fei_hevc.h]: could you replace Hevc with HEVC for all data structure definitions in this file?  We use HEVC in other data structures, please keep the same name style.

File: va/va_fei_hevc.h [hunk 3]
```diff
@@ -0,0 +1,282 @@
...
+
+    /** \brief number of MV predictors L0 and L1. the maximum number of motion vector predictor for a 16x16, 32x32 or
+     * 64x64 block is four, it must not be greater than maximum supported MV predictor,
+     **/
+    uint32_t      num_mv_predictors_l0      : 16;
+    uint32_t      num_mv_predictors_l1      : 16;
+
+    /** \brief control parameters */
+    uint32_t      search_path               : 8;
+    uint32_t      len_sp                    : 8;
```
Comment (xhaihao) [inline va/va_fei_hevc.h]: what does 'sp' stand for? could you add more comment for this field or use a meaningful name?

File: va/va_fei_hevc.h [hunk 4]
```diff
@@ -0,0 +1,282 @@
...
+ * it is output buffer of ENC and ENC_PAK modes, it's also input buffer of PAK mode.
+ * The CTB block is in raster scan order. Buffer size shall not be less than the number of CTB blocks
+ * multiplied by sizeof(VAEncFEICTBCmdHevc). Note that, when separate ENC and PAK is enabled,
+ * the exact layout of this buffer is needed for PAK input. App can reuse this buffer,
+ * or copy to a different buffer as PAK input, reserved elements must not be modified when used as PAK input.
+ **/
+typedef struct _VAEncFEICTBCmdHevc
+{
+    // DWORD 0
+    uint32_t    reserved0;
```
Comment (xhaihao) [inline va/va_fei_hevc.h]: Could you move the reserved bytes to the end of this structure?

File: va/va_fei_hevc.h [hunk 5]
```diff
@@ -0,0 +1,282 @@
...
+    uint32_t    split_flag_level1                : 4;
+    uint32_t    split_flag_level0                : 1;
+    uint32_t    reserved1                        : 3;
+    uint32_t    cu_count_minus1                  : 6;
+    uint32_t    last_ctb_of_tile_flag            : 1;
+    uint32_t    last_ctb_of_slice_flag           : 1;
+
+    // DWORD 2
+    uint32_t    current_lcu_Xaddr                : 16;
+    uint32_t    current_lcu_Yaddr                : 16;
```
Comment (xhaihao) [inline va/va_fei_hevc.h]: could you use lowercase for each field?

File: va/va_fei_hevc.h (hunk 3 above)
Comment (kelvinhu325) [reply va/va_fei_hevc.h]: SP means Search Path, len_sp means Length of Search Path.

File: va/va_fei_hevc.h (hunk 4 above)
Comment (kelvinhu325) [reply va/va_fei_hevc.h]: No, this struct is HW PAK engine input layout, we can not move reserved0, otherwise HW PAK may gets hang.

Comment (xhaihao) [reply va/va_fei_hevc.h]: so this buffer is hardware dependent, right?

Comment (kelvinhu325) [reply va/va_fei_hevc.h]: yes, actually the first DWORD probable includes some debug info , which does not want customers know, it does not mean only a reserved DWORD.

Comment (kelvinhu325) [reply va/va_fei_hevc.h]: Yes

**Comment by sreerenjb** [comment]
Many things are unclear in this PR
-- it seem the va.h is missing the buffer type definitions of HEVC FEI.
-- I didn't see the definition of VAEncFEICURecordLongHevc
-- no defintion of VAEncFEICURecordHevc
-- no definitin of VAEncFEIDistortionBufferHevc
-- if we are going to put comment like this " Each CTB block has one CU record data with layout VAEncFEICURecordLongHevc for SKL and VAEncFEICURecordHevc for ICL+" then it is platform specific apis, we should move them under vendor/intel or something similar.

**Pull request merged**